### PR TITLE
perf: bookings query improvement

### PR DESCRIPTION
**Pull request by hariombalhara**
## bookings.get handler
In a DB with ~15000 bookings and ~150K even-types, for a user having 900 bookings it takes on average 100ms locally for the bookings.get handler to respond with `main` branch.
In this branch, it takes on average 30ms for the same
I expect improvements to be much better in production with more data and latency b/w DB and servers.

Also added a seeder for performance testing locally(seed-performance-testing.ts). Courtesy of @emrysal 
## Type of change
- [x] Perf Improvement

## How should this be tested?
- Go to /bookings/upcoming and measure the response time of bookings/get endpoint
- Regression Test
	- Test filters with different event-types and users
	- Test that when we do load more of bookings, there are no duplicate bookings

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/lib/server/addServerTimingHeader.ts
Comment (hariombalhara) [inline packages/lib/server/addServerTimingHeader.ts:1]: To be replaced with otel in branch https://github.com/calcom/cal.com/pull/10574/files#diff-3bcd8b09138a03622b300b58836dfed0598874aed41ff92f57c2b2700f30f4b5

File: packages/prisma/seed-performance-testing.ts
Comment (hariombalhara) [inline packages/prisma/seed-performance-testing.ts:1]: Adding it to make it easy for anyone to seed the DB for performance testing.
It has two functions right now. 
1. Creates many users with many events and their bookings
2. Creates a single user with many many bookings.

File: apps/web/pages/_document.tsx [hunk 1]
```diff
@@ -49,7 +49,7 @@ class MyDocument extends Document<Props> {
           <meta name="msapplication-TileColor" content="#ff0000" />
           <meta name="theme-color" media="(prefers-color-scheme: light)" content="#f9fafb" />
           <meta name="theme-color" media="(prefers-color-scheme: dark)" content="#1C1C1C" />
-          {(process.env.NODE_ENV === "development" || process.env.VERCEL_ENV === "preview") && (
+          {process.env.VERCEL_ENV === "preview" && (
```
Comment (hariombalhara) [inline apps/web/pages/_document.tsx:52]: We don't need meticulous on local, it causes unnecessary requests.

File: packages/trpc/server/createNextApiHandler.ts [hunk 2]
```diff
@@ -15,7 +16,20 @@ export function createNextApiHandler(router: AnyRouter, isPublic = false, namesp
...
+      addServerTimingHeaderIfRes({
+        res,
+        timings: [
+          {
+            label: "createContext",
+            duration: end - start,
+          },
+        ],
+      });
+      return context;
```
Comment (hariombalhara) [inline packages/trpc/server/createNextApiHandler.ts:32]:
```suggestion
      return createTrpcContext({ req, res });
```

File: packages/trpc/server/createNextApiHandler.ts [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import { z } from "zod";
 
+import { addServerTimingHeaderIfRes } from "@calcom/lib/server/addServerTimingHeader";
```
Comment (hariombalhara) [inline packages/trpc/server/createNextApiHandler.ts:3]:
```suggestion
```

File: packages/trpc/server/middlewares/localeMiddleware.ts [hunk 4]
```diff
@@ -10,6 +13,17 @@ const localeMiddleware = sessionMiddleware.unstable_pipe(async ({ ctx, next }) =
   );
 
   const locale = user?.locale || ctx.locale;
+  const middlewareEnd = performance.now();
+
+  addServerTimingHeaderIfRes({
```
Comment (hariombalhara) [inline packages/trpc/server/middlewares/localeMiddleware.ts:18]: To be replaced with otel api

File: packages/trpc/server/middlewares/sessionMiddleware.ts [hunk 5]
```diff
@@ -148,17 +149,31 @@ const getUserSession = async (ctx: TRPCContextInner) => {
 
   return { user, session };
 };
+
 const sessionMiddleware = middleware(async ({ ctx, next }) => {
+  const middlewareStart = performance.now();
   const { user, session } = await getUserSession(ctx);
-
+  const middlewareEnd = performance.now();
+  addServerTimingHeaderIfRes({
```
Comment (hariombalhara) [inline packages/trpc/server/middlewares/sessionMiddleware.ts:157]: to be replaed

File: packages/app-store/routing-forms/trpc/deleteForm.handler.ts [hunk 6]
```diff
@@ -1,6 +1,5 @@
-import type { PrismaClient } from "@prisma/client";
-
 import { entityPrismaWhereClause } from "@calcom/lib/entityPermissionUtils";
+import type { PrismaType } from "@calcom/prisma";
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/trpc/deleteForm.handler.ts:2]: Made these changes to support Prisma Client extensions. We anyway should have read the prisma type from @calcom/prisma only.

File: packages/prisma/index.ts [hunk 7]
```diff
@@ -1,28 +1,65 @@
...
 
 export const customPrisma = (options: Prisma.PrismaClientOptions) =>
-  new PrismaClient({ ...prismaOptions, ...options });
+  new PrismaClientWithoutExtension({ ...prismaOptions, ...options });
+
+// If any changed on middleware server restart is required
+// TODO: Migrate it to $extends
+bookingReferenceMiddleware(prismaWithoutClientExtensions);
+
+// FIXME: Due to some reason, there are types failing in certain places due to the $extends. Fix it and then enable it
```
Comment (hariombalhara) [inline packages/prisma/index.ts:24]: Will resolve this problem once this PR is merged.

File: packages/trpc/server/middlewares/perfMiddleware.ts [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import { middleware } from "../trpc";
 
-const perfMiddleware = middleware(async ({ path, type, next }) => {
+const perfMiddleware = middleware(async ({ path, type, next, ctx }) => {
```
Comment (hariombalhara) [inline packages/trpc/server/middlewares/perfMiddleware.ts:3]:
```suggestion
const perfMiddleware = middleware(async ({ path, type, next }) => {
```

File: packages/trpc/server/routers/viewer/bookings/get.handler.ts [hunk 9]
```diff
@@ -72,6 +73,62 @@ export const getHandler = async ({ ctx, input }: GetOptions) => {
...
+const set = new Set();
+const getUniqueBookings = <T extends { uid: string }>(arr: T[]) => {
+  const unique = arr.filter((booking) => {
+    const duplicate = set.has(booking.uid);
+    set.add(booking.uid);
+    return !duplicate;
+  });
+  set.clear();
+  return unique;
+};
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/bookings/get.handler.ts:113]: This is existing code, just being shown like this due to movement of a lot of code.

File: packages/trpc/server/routers/viewer/bookings/get.handler.ts [hunk 10]
```diff
@@ -162,65 +318,6 @@ export const getHandler = async ({ ctx, input }: GetOptions) => {
...
-          select: {
-            referenceUid: true,
-            attendee: {
-              select: {
-                email: true,
-              },
-            },
-          },
-        },
-      },
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/bookings/get.handler.ts:223]: Moved it to `bookingSelect` as it's shared across all queries.

File: packages/trpc/server/routers/viewer/bookings/get.handler.ts [hunk 11]
```diff
@@ -106,36 +163,125 @@ export const getHandler = async ({ ctx, input }: GetOptions) => {
...
+  };
 
-  const [bookingsQuery, recurringInfoBasic, recurringInfoExtended] = await Promise.all([
+  const [
+    // Quering these in parallel to save time.
+    // Note that because we are applying `take` to individual queries, we will usually get more bookings then we need. It is okay to have more bookings faster than having what we need slower
+    bookingsQueryUserId,
+    bookingsQueryAttendees,
+    bookingsQueryTeamMember,
+    bookingsQuerySeatReference,
```
Comment (hariombalhara) [inline packages/trpc/server/routers/viewer/bookings/get.handler.ts:245]: Most important piece to review

**Review by keithwillcode — APPROVED** [review]
Nice refactor.

File: packages/prisma/index.ts [hunk 12]
```diff
@@ -1,28 +1,65 @@
...
-  var prisma: PrismaClient | undefined;
+  var prisma: typeof prismaWithClientExtensions;
 }
 
 const prismaOptions: Prisma.PrismaClientOptions = {};
 
 if (!!process.env.NEXT_PUBLIC_DEBUG) prismaOptions.log = ["query", "error", "warn"];
 
-export const prisma = globalThis.prisma || new PrismaClient(prismaOptions);
+const prismaWithoutClientExtensions = new PrismaClientWithoutExtension(prismaOptions);
```
Comment (zomars) [inline packages/prisma/index.ts:15]: Seems like this change is the one causing a lot of idle instances:

```
warn(prisma-client) This is the 10th instance of Prisma Client being started. Make sure this is intentional
```

Comment (zomars) [reply packages/prisma/index.ts:15]: cc @emrysal

**Pull request merged**